### PR TITLE
Remove union-types

### PR DESCRIPTION
**Pull request by FredTheDino**

### CODE REVIEW

File: sylt-parser/src/parser.rs [hunk 1]
```diff
@@ -673,29 +671,6 @@ pub fn parse_type<'t>(ctx: Context<'t>) -> ParseResult<'t, Type> {
     // Wrap it in a syntax tree node.
     let ty = Type { span, kind };
```
Comment (EMattfolk) [inline sylt-parser/src/parser.rs:673]:
```suggestion
```
This could be inlined further down.

File: sylt-std/src/sylt.rs [hunk 2]
```diff
@@ -79,7 +79,7 @@ sylt_macro::extern_function!(
     "sylt_std::sylt",
     reduce,
     ? "Reduce the list to a single element, returns 'nil' if the input list is empty",
-    -> "fn [*ITEM], fn *ITEM, *ITEM -> *OUT -> *OUT | nil",
+    -> "fn [*ITEM], (fn *ITEM, *ITEM -> *OUT) -> *OUT", // The return is a lie!
```
Comment (EMattfolk) [inline sylt-std/src/sylt.rs:82]: Yes, this is why reduce is not as nice as fold.

File: sylt-std/src/sylt.rs [hunk 3]
```diff
@@ -183,10 +183,13 @@ sylt_macro::extern_function!(
     "sylt_std::sylt",
     len,
     ? "Gives the length of list",
-    -> "fn [*ITEM] | {*KEY: *VALUE} -> int",
+    -> "fn<C: Container> *C -> int",
```
Comment (EMattfolk) [inline sylt-std/src/sylt.rs:186]: :open_mouth: This does look pretty sick.

File: sylt-std/src/sylt.rs [hunk 4]
```diff
@@ -326,7 +329,7 @@ sylt_macro::extern_function!(
     "sylt_std::sylt",
     sign,
     ? "Returns the sign of the value",
-    -> "(fn int -> int) | (fn float -> float)",
+    -> "fn<N: Num> *N -> *N",
```
Comment (EMattfolk) [inline sylt-std/src/sylt.rs:332]: Finally!

File: sylt-std/src/sylt.rs [hunk 5]
```diff
@@ -456,7 +461,7 @@ sylt_macro::extern_function!(
     "sylt_std::sylt",
     dot,
     ? "Computes the scalar product",
-    -> "fn (float, float) -> float",
+    -> "fn (float, float), (float, float) -> float",
```
Comment (EMattfolk) [inline sylt-std/src/sylt.rs:464]: This one could be a num as well. But no need to change it now.

**Review by EMattfolk — APPROVED** [review]

**Review by sornas — COMMENTED** [review]
looks great

File: sylt-compiler/src/typechecker.rs [hunk 6]
```diff
@@ -268,7 +268,16 @@ impl TypeChecker {
...
-                _ => return err_type_error!(self, span, ctx, todo_error!()),
+                _ => {
+                    return err_type_error!(
+                        self,
+                        ident.span,
+                        ctx,
+                        TypeError::Exotic,
+                        "Cannot find type '{}' - is it perhaps a type-variable?",
+                        ident.name
+                    )
```
[the remark this answers is not in the document]
Comment (FredTheDino) [reply sylt-compiler/src/typechecker.rs:279]: Yes, why i added it. If you write a type like: `fn C -> *`
If you forget a type variable or the type isn't defined. Should probably not the exotic.

File: sylt-parser/src/parser.rs (hunk 1 above)
Comment (FredTheDino) [reply sylt-parser/src/parser.rs:673]: Go for it

File: sylt-std/src/sylt.rs (hunk 5 above)
Comment (FredTheDino) [reply sylt-std/src/sylt.rs:464]: We'd need to change the function, the break up vec function returns floats.

**Pull request merged**